src/v/storage/index_state.cc

```diff
@@ -265,14 +266,15 @@ bool index_state::maybe_index(
 
     // index_state
     bool is_empty = false;
+    bool should_set_non_data_timestamp = false;
     if (empty()) {
         // Ordinarily, we do not allow configuration batches to contribute to
         // the segment's timestamp bounds (because config batches use walltime
         // but user data timestamps may be anything).  However, for the first
         // batch we set the timestamps, and then set a `non_data_timestamps`
         // flag so that the next time we see user data we will overwrite
         // the walltime timestamps with the user data timestamps.
-        non_data_timestamps = !user_data;
+        should_set_non_data_timestamp = !user_data;
```

Comment:
Note for other reviewers, i found it easier to reason about this PR calling this `first_entry_is_non_data`